File: tools/parity/probe.ts
import { parseOrderedJson, pythonDumps, PythonFloat, type OrderedJson } from "./serialize";
import { unicode15Casefold } from "./casefold";

type PlainCase = { id: string; kind: string; input: any; mitigation: string };
type OrderedCase = Map<string, OrderedJson>;

const casesUrl = new URL("./cases.json", import.meta.url);
const source = await Bun.file(casesUrl).text();
const plainCases = JSON.parse(source).cases as PlainCase[];
const orderedRoot = parseOrderedJson(source) as Map<string, OrderedJson>;
const orderedCases = orderedRoot.get("cases") as OrderedCase[];

function mapGet<T extends OrderedJson>(map: Map<string, OrderedJson>, key: string): T {
  return map.get(key) as T;
}

function naiveRegex(spec: { pattern: string; text: string }): string {
  try {
    const regex = new RegExp(spec.pattern, "gm");
    return JSON.stringify([...spec.text.matchAll(regex)].map((match) => [...match]));
  } catch {
    // 컴파일 실패도 regex 답의 배열 형식을 유지해야 corpus 전체를 끝까지 비교할 수 있다.
    return JSON.stringify([]);
  }
}

function naiveAnswer(test: PlainCase): string {
  const value = test.input;
  switch (test.kind) {
    case "json": return JSON.stringify(value);
    case "json_pretty": return JSON.stringify(value, null, 2);
    case "sha256": return new Bun.CryptoHasher("sha256").update(String(value)).digest("hex");
    case "nfc": return [...String(value).normalize("NFC")]
      .map((char) => `U+${char.codePointAt(0)!.toString(16).toUpperCase().padStart(4, "0")}`).join(" ");
    case "casefold": return JSON.stringify(value.map((item: unknown) => String(item).toLowerCase()));
    case "sort": return JSON.stringify(value.map(String).sort());
    case "codepoints": {
      const text = String(value);
      return JSON.stringify({ length: text.length, head2: text.slice(0, 2) });
    }
    case "regex": return naiveRegex(value);
    default: throw new Error(`unknown case kind: ${test.kind}`);
  }
}

function compareCodePoints(left: string, right: string): number {
  const a = [...left].map((char) => char.codePointAt(0)!);
  const b = [...right].map((char) => char.codePointAt(0)!);
  for (let index = 0; index < Math.min(a.length, b.length); index++) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

function pythonRegex(pattern: string): RegExp {
  let flags = "gm";
  const inline = pattern.match(/^\(\?([aiLmsux]+)\)/);
  if (inline) {
    pattern = pattern.slice(inline[0].length);
    if (inline[1].includes("i")) flags += "i";
    if (inline[1].includes("s")) flags += "s";
  }
  // Python의 \d는 Unicode Decimal_Number이므로 JS의 ASCII 전용 escape를 그대로 쓰면 안 된다.
  pattern = pattern.replace(/(^|[^\\])\\d/g, "$1\\p{Decimal_Number}");
  if (pattern.includes("\\p{Decimal_Number}") && !flags.includes("u")) flags += "u";
  return new RegExp(pattern, [...new Set(flags)].join(""));
}

function mitigatedRegex(spec: Map<string, OrderedJson>): string {
  const pattern = mapGet<string>(spec, "pattern");
  const text = mapGet<string>(spec, "text");
  const matches = [...text.matchAll(pythonRegex(pattern))]
    .map((match) => [...match].map((group) => group === undefined ? null : group));
  return pythonDumps(matches);
}

function primitiveString(value: OrderedJson): string {
  if (value instanceof PythonFloat) return String(value.value);
  return String(value);
}

function mitigatedAnswer(test: OrderedCase): string {
  const kind = mapGet<string>(test, "kind");
  const value = mapGet<OrderedJson>(test, "input");
  switch (kind) {
    case "json": return pythonDumps(value);
    case "json_pretty": return pythonDumps(value, true);
    case "sha256": return new Bun.CryptoHasher("sha256").update(primitiveString(value)).digest("hex");
    case "nfc": return [...primitiveString(value).normalize("NFC")]
      .map((char) => `U+${char.codePointAt(0)!.toString(16).toUpperCase().padStart(4, "0")}`).join(" ");
    case "casefold": return pythonDumps((value as OrderedJson[]).map((item) => unicode15Casefold(primitiveString(item))));
    case "sort": return pythonDumps((value as OrderedJson[]).map(primitiveString).sort(compareCodePoints));
    case "codepoints": {
      const points = [...primitiveString(value)];
      return pythonDumps(new Map<string, OrderedJson>([["length", points.length], ["head2", points.slice(0, 2).join("")]]));
    }
    case "regex": return mitigatedRegex(value as Map<string, OrderedJson>);
    default: throw new Error(`unknown case kind: ${kind}`);
  }
}

const results = Object.fromEntries(plainCases.map((test) => [test.id, naiveAnswer(test)]));
const mitigated = Object.fromEntries(orderedCases.map((test) => [mapGet<string>(test, "id"), mitigatedAnswer(test)]));
const versions = process.versions as Record<string, string | undefined>;

process.stdout.write(JSON.stringify({
  runtime: `bun ${Bun.version}`,
  ...(versions.unicode ? { unicode: versions.unicode } : {}),
  results,
  mitigated,
}, null, 2) + "\n");

File: tools/parity/probe.py
#!/usr/bin/env python3
"""Parity corpus — Python(정본) 쪽 답안지.

`tools/parity/cases.json` 을 읽어 case 마다 한 줄짜리 문자열 답을 낸다.
`probe.ts` 가 같은 입력으로 같은 모양의 답을 내고, `parity.py corpus` 가 둘을
대조한다. 어느 쪽도 상대의 출력을 입력으로 쓰지 않는다 — 같은 원본에서 독립
실행해야 대조가 의미를 가진다.

출력: {"runtime": "...", "results": {case_id: answer}}
"""
from __future__ import annotations

import hashlib
import json
import re
import sys
import unicodedata
from pathlib import Path
from typing import Any

CASES = Path(__file__).resolve().parent / "cases.json"
# Python 정규식의 inline flag. JS 에는 없어서 분리해야 하는 지점을 그대로 드러낸다.
INLINE_FLAG = re.compile(r"^\(\?([aiLmsux]+)\)")


def dumps(value: Any, *, pretty: bool) -> str:
    """scripts/llmwiki.py 의 dump() 와 같은 규칙."""
    return json.dumps(value, ensure_ascii=False, indent=2 if pretty else None,
                      separators=None if pretty else (",", ":"))


def run_regex(spec: dict[str, Any]) -> str:
    pattern, text = spec["pattern"], spec["text"]
    flags = 0
    match = INLINE_FLAG.match(pattern)
    if match:
        for letter in match.group(1):
            flags |= {"i": re.I, "m": re.M, "s": re.S, "x": re.X, "a": re.A}.get(letter, 0)
        pattern = pattern[match.end():]
    found = [[group for group in ([m.group(0)] + list(m.groups()))]
             for m in re.finditer(pattern, text, flags | re.M)]
    return dumps(found, pretty=False)


def answer(case: dict[str, Any]) -> str:
    kind, value = case["kind"], case["input"]
    if kind == "json":
        return dumps(value, pretty=False)
    if kind == "json_pretty":
        return dumps(value, pretty=True)
    if kind == "sha256":
        return hashlib.sha256(str(value).encode("utf-8")).hexdigest()
    if kind == "nfc":
        normalized = unicodedata.normalize("NFC", str(value))
        return " ".join(f"U+{ord(ch):04X}" for ch in normalized)
    if kind == "casefold":
        return dumps([str(item).casefold() for item in value], pretty=False)
    if kind == "sort":
        return dumps(sorted(str(item) for item in value), pretty=False)
    if kind == "codepoints":
        text = str(value)
        return dumps({"length": len(text), "head2": text[:2]}, pretty=False)
    if kind == "regex":
        return run_regex(value)
    raise SystemExit(f"unknown case kind: {kind}")


def main() -> None:
    cases = json.loads(CASES.read_text(encoding="utf-8"))["cases"]
    results = {case["id"]: answer(case) for case in cases}
    sys.stdout.write(json.dumps(
        {"runtime": f"python {sys.version_info.major}.{sys.version_info.minor}."
                    f"{sys.version_info.micro}",
         "unicode": unicodedata.unidata_version, "results": results},
        ensure_ascii=False, indent=2) + "\n")


if __name__ == "__main__":
    main()
